lib/src/layer/polygon_layer.dart

```diff
@@ -58,133 +72,190 @@ class PolygonLayer extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
-    return LayoutBuilder(
-      builder: (BuildContext context, BoxConstraints bc) {
-        final map = FlutterMapState.maybeOf(context)!;
-        final size = Size(bc.maxWidth, bc.maxHeight);
-        final polygonsWidget = <Widget>[];
-
-        for (final polygon in polygons) {
-          polygon.offsets.clear();
-
-          if (null != polygon.holeOffsetsList) {
-            for (final offsets in polygon.holeOffsetsList!) {
-              offsets.clear();
-            }
-          }
-
-          if (polygonCulling &&
-              !polygon.boundingBox.isOverlapping(map.bounds)) {
-            // skip this polygon as it's offscreen
-            continue;
-          }
-
-          _fillOffsets(polygon.offsets, polygon.points, map);
-
-          if (null != polygon.holePointsList) {
-            final len = polygon.holePointsList!.length;
-            for (var i = 0; i < len; ++i) {
-              _fillOffsets(
-                  polygon.holeOffsetsList![i], polygon.holePointsList![i], map);
-            }
-          }
-
-          polygonsWidget.add(
-            CustomPaint(
-              key: polygon.key,
-              painter: PolygonPainter(polygon, map.rotationRad),
-              size: size,
-            ),
-          );
-        }
-
-        return Stack(
-          children: polygonsWidget,
-        );
-      },
+    final map = FlutterMapState.maybeOf(context)!;
+    final size = Size(map.size.x, map.size.y);
+    final origin = map.pixelOrigin;
+    final offset = Offset(origin.x.toDouble(), origin.y.toDouble());
+
+    final Iterable<Polygon> pgons = polygonCulling
+        ? polygons.where((p) {
+            return p.boundingBox.isOverlapping(map.bounds);
+          })
+        : polygons;
+
+    final paint = CustomPaint(
+      painter: PolygonPainter(pgons, map),
+      size: size,
+      willChange: false,
+      isComplex: true,
+    );
+    return Positioned(
+      left: -offset.dx,
+      top: -offset.dy,
+      child: kIsWeb ? paint : RepaintBoundary(child: paint),
     );
-  }
-
-  void _fillOffsets(final List<Offset> offsets, final List<LatLng> points,
-      FlutterMapState map) {
-    final len = points.length;
-    for (var i = 0; i < len; ++i) {
-      final point = points[i];
-      final offset = map.getOffsetFromOrigin(point);
-      offsets.add(offset);
-    }
   }
 }
 
 class PolygonPainter extends CustomPainter {
-  final Polygon polygonOpt;
-  final double rotationRad;
+  final Iterable<Polygon> polygons;
+  final FlutterMapState map;
+  final double zoom;
+  final double rotation;
+
+  PolygonPainter(this.polygons, this.map)
+      : zoom = map.zoom,
+        rotation = map.rotation;
+
+  int get hash {
+    _hash ??= Object.hashAll(polygons);
+    return _hash!;
+  }
+
+  int? _hash;
 
-  PolygonPainter(this.polygonOpt, this.rotationRad);
+  List<Offset> getOffsets(List<LatLng> points) {
+    return points.map((pos) => getOffset(pos)).toList();
```

Comment:
```suggestion
    return List.generate(points.length, (index) {
      final delta = map.project(points[index]);
      return Offset(delta.x.toDouble(), delta.y.toDouble());
    }, growable: false);
```

This is significantly more performant since it generates the list directly. I saw 8ms -> 0.4ms in profiling 500 polygons

Reply:
Nice!